mockos/lvm.go

```diff
@@ -0,0 +1,291 @@
+package mockos
+
+import (
+	"fmt"
+	"path"
+
+	"github.com/anuvu/disko"
+)
+
+type mockLVM struct {
+	VGs     disko.VGSet `json:"vgs"`
+	PVs     disko.PVSet `json:"pvs"`
+	sys     disko.System
+	freePVs disko.PVSet
+}
+
+// LVM return mock lvm implementation.
+func LVM(sys disko.System) disko.VolumeManager {
+	return &mockLVM{
+		VGs:     disko.VGSet{},
+		PVs:     disko.PVSet{},
+		sys:     sys,
+		freePVs: disko.PVSet{},
+	}
+}
+
+func (lvm *mockLVM) ScanPVs(filter disko.PVFilter) (disko.PVSet, error) {
+	pvs := disko.PVSet{}
+
+	for n, pv := range lvm.PVs {
+		if filter == nil || filter(pv) {
+			pvs[n] = pv
+		}
+	}
+
+	return pvs, nil
+}
+
+func (lvm *mockLVM) ScanVGs(filter disko.VGFilter) (disko.VGSet, error) {
+	vgs := disko.VGSet{}
+
+	for n, vg := range lvm.VGs {
+		if filter == nil || filter(vg) {
+			vgs[n] = vg
+		}
+	}
+
+	return vgs, nil
+}
+
+func hasPartition(disks disko.DiskSet, name string) bool {
+	for _, d := range disks {
+		for _, p := range d.Partitions {
+			if p.Name == name {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (lvm *mockLVM) CreatePV(diskName string) (disko.PV, error) {
+	disks, _ := lvm.sys.ScanAllDisks(func(d disko.Disk) bool { return true })
+	d, ok := disks[diskName]
+
+	if !ok {
+		if !hasPartition(disks, diskName) {
+			return disko.PV{}, fmt.Errorf("disk %s does not exist", diskName)
+		}
+	}
+
+	if _, ok := lvm.PVs[diskName]; ok {
+		return disko.PV{}, fmt.Errorf("pv %s already exists", diskName)
+	}
+
+	pv := disko.PV{
+		Name:     diskName,
+		Path:     path.Join("dev", diskName),
+		Size:     d.Size,
+		FreeSize: d.Size,
+	}
+
+	lvm.freePVs[pv.Name] = pv
+	lvm.PVs[pv.Name] = pv
+
+	return pv, nil
+}
+
+func (lvm *mockLVM) DeletePV(pv disko.PV) error {
+	if _, ok := lvm.PVs[pv.Name]; !ok {
+		return fmt.Errorf("pv %s does not exist", pv.Name)
+	}
+
+	// PV must not be used by any vg to delete
+	if _, ok := lvm.freePVs[pv.Name]; !ok {
+		return fmt.Errorf("pv %s is in use", pv.Name)
+	}
+
+	delete(lvm.PVs, pv.Name)
+	delete(lvm.freePVs, pv.Name)
+
+	return nil
+}
+
+func (lvm *mockLVM) HasPV(name string) bool {
```

Comment:
what is the name here?  A full path?  just a basename (sda)
What is the correct usage of HasPV() ?
Assuming `HasPV("sda")` is correct, then what does does HasPV("sda") answer?

   a.) Does /dev/sda exist as a block device?
   b.) Does /dev/sda contain a lvm pv device header already (and has been 
   c.) Is /dev/sda a member of any VolumeGroups

The implementation Would seem to indicate 'b'.

Reply:
Yes. I was assuming #b. I can add a comment to the method.

Reply:
Name by convention across all of the api is just the device kname "sda", Path is what we are using as the full path "/dev/sda".